unit/java_bytecode/java_bytecode_parse_generics/parse_bounded_generic_inner_classes.cpp

```diff
@@ -0,0 +1,231 @@
+    REQUIRE(is_java_generic_type(belem_type));
+    THEN("Field has instantiated type variable")
+    {
+      const java_generic_typet &container = to_java_generic_type(belem_type);
+
+      const std::vector<java_generic_parametert> &generic_types =
+        container.generic_type_variables();
+      REQUIRE(generic_types.size() == 1);
+
+      const typet &inst_type = java_generic_get_inst_type(0, container);
+
+      REQUIRE(inst_type.id() == ID_pointer);
+      const typet &inst_type_symbol = inst_type.subtype();
+      REQUIRE(inst_type_symbol.id() == ID_symbol);
+      REQUIRE(
+        to_symbol_type(inst_type_symbol).get_identifier() ==
+        "java::java.lang.Integer");
+    }
+  }
+
+  WHEN("Parsing an inner class with double bounded type variable")
+  {
+    std::string doubleboundedinner_name = class_prefix + "$DoubleBoundedInner";
+    REQUIRE(new_symbol_table.has_symbol(doubleboundedinner_name));
+    THEN("The bounds should be encoded")
+    {
+      const symbolt &class_symbol =
+        new_symbol_table.lookup_ref(doubleboundedinner_name);
+      class_typet class_type =
+        require_symbol::require_complete_class(class_symbol);
+
+      java_class_typet java_class_type = to_java_class_type(class_type);
+      REQUIRE_FALSE(is_java_generics_class_type(java_class_type));
+
+      // TODO (tkiley): Extend this unit test when bounds are correctly
+      // parsed.
+#if 0
+      java_generics_class_typet java_generics_class_type=
+        to_java_generics_class_type(java_class_type);
+      REQUIRE(java_generics_class_type.generic_types().size()==1);
+      typet &type_var=java_generics_class_type.generic_types().front();
+      REQUIRE(is_java_generic_parameter(type_var));
+      java_generic_parametert generic_type_var=
+        to_java_generic_parameter(type_var);
+
+      REQUIRE(
+        generic_type_var.type_variable().get_identifier()==
+        doubleboundedinner_name+"::T");
+      REQUIRE(
+        java_generics_class_type_var(0, java_generics_class_type)==
+        doubleboundedinner_name+"::T");
+      THEN("Bound must be Number and Interface")
+      {
+
+      }
+#endif
+    }
+  }
+
+  GIVEN("An inner class with multiple generic parameters")
+  {
+    std::string twoelementinner_name = class_prefix + "$TwoElementInner";
+    REQUIRE(new_symbol_table.has_symbol(twoelementinner_name));
+    THEN("Both generic parameters should be encoded")
+    {
+      const symbolt &class_symbol =
+        new_symbol_table.lookup_ref(twoelementinner_name);
+      class_typet class_type =
+        require_symbol::require_complete_class(class_symbol);
+      java_generics_class_typet java_generics_class_type =
+        require_type::require_java_generic_class(class_type);
+
+      REQUIRE(java_generics_class_type.generic_types().size() == 2);
+
+      auto generic_param_iterator =
+        java_generics_class_type.generic_types().cbegin();
+
+      // The first parameter should be called K
```

Comment:
There seems to be a couple of places where we do things like 'first type parameter should be called K' - could the code for that be factored out into a function in `test_utils/require_type` or somewhere similar?